backend-rust/src/graphql_api/transaction_metrics.rs

```diff
@@ -0,0 +1,119 @@
+    /// Total number of transactions (all time).
+    last_cumulative_transaction_count: i64,
+    /// Total number of transactions in the requested period.
+    transaction_count: i64,
+    buckets: TransactionMetricsBuckets,
+}
+
+#[derive(SimpleObject)]
+struct TransactionMetricsBuckets {
+    /// The width (time interval) of each bucket.
+    bucket_width: TimeSpan,
+
+    /// Start of the bucket time period. Intended x-axis value.
+    #[graphql(name = "x_Time")]
+    x_time: Vec<DateTime>,
+
+    /// Total number of transactions (all time) at the end of the bucket period.
+    /// Intended y-axis value.
+    #[graphql(name = "y_LastCumulativeTransactionCount")]
+    y_last_cumulative_transaction_count: Vec<i64>,
+
+    /// Total number of transactions within the bucket time period. Intended
+    /// y-axis value.
+    #[graphql(name = "y_TransactionCount")]
+    y_transaction_count: Vec<i64>,
+}
+
+#[Object]
+impl TransactionMetricsQuery {
+    async fn transaction_metrics(
+        &self,
+        ctx: &Context<'_>,
+        period: MetricsPeriod,
+    ) -> ApiResult<TransactionMetrics> {
+        let pool = get_pool(ctx)?;
+
+        let last_cumulative_transaction_count = sqlx::query_scalar!(
+            "SELECT cumulative_num_txs FROM blocks ORDER BY height DESC LIMIT 1"
+        )
+        .fetch_one(pool)
+        .await?;
+
+        let interval: PgInterval = period
+            .as_duration()
+            .try_into()
+            .map_err(|e| ApiError::DurationOutOfRange(Arc::new(e)))?;
+
+        let cumulative_transaction_count_before_period = sqlx::query_scalar!(
+            "SELECT cumulative_num_txs
+            FROM blocks
+            WHERE slot_time < (now() - $1::interval)
+            ORDER BY height DESC
+            LIMIT 1",
+            interval,
+        )
+        .fetch_one(pool)
+        .await?;
+
+        let transaction_count =
+            last_cumulative_transaction_count - cumulative_transaction_count_before_period;
+
+        let bucket_width = period.bucket_width();
+
+        // The full period interval, e.g. 7 days.
+        let period_interval: PgInterval = period
+            .as_duration()
+            .try_into()
+            .map_err(|err| ApiError::DurationOutOfRange(Arc::new(err)))?;
+
+        // The bucket interval, e.g. 6 hours.
+        let bucket_interval: PgInterval =
+            bucket_width.try_into().map_err(|err| ApiError::DurationOutOfRange(Arc::new(err)))?;
+
+        let rows = sqlx::query_file!(
+            "src/graphql_api/transaction_metrics.sql",
+            period_interval,
+            bucket_interval,
+        )
+        .fetch_all(pool)
+        .await?;
+
+        let x_time = rows
+            .iter()
+            .map(|r| r.bucket_time.expect("generated by generate_series so never null"))
+            .collect();
+        let y_last_cumulative_transaction_count =
+            rows.iter().map(|r| r.end_cumulative_num_txs.expect("coalesced")).collect();
```

Comment:
Suggestion: An additional explanation in the `expect` part would help to distinguish the error from similar errors further down. E.g. I would mention the field that caused the error: `end_cumulative_num_txs`

Reply:
First of all, the panic message would give the line number so it would be distinguished from the others.

But secondly, the explanation of the expect here is meant to say "this will never panic because this field is coalesced, i.e. using the `COALESCE` function in SQL, therefore it is never `NULL`, therefore this is always `Some`". So the error will never happen in the first place.